crates/solvers/src/boundary/rate_limiter.rs

```diff
@@ -0,0 +1,219 @@
+use {
+    anyhow::{ensure, Context, Result},
+    shared::rate_limiter::{
+        RateLimiter as SharedRateLimiter,
+        RateLimiterError as SharedRateLimiterError,
+        RateLimitingStrategy as SharedRateLimitingStrategy,
+    },
+    std::{future::Future, ops::Add, str::FromStr, time::Duration},
+    thiserror::Error,
+};
+
+pub struct RateLimiter {
+    inner: SharedRateLimiter,
+    max_retries: usize,
+}
+
+#[derive(Debug, Clone)]
+pub struct RateLimitingStrategy {
+    inner: SharedRateLimitingStrategy,
+    max_retries: usize,
+}
+
+const DEFAULT_MAX_RETIRES: usize = 1;
+
+impl Default for RateLimitingStrategy {
+    fn default() -> Self {
+        Self {
+            inner: SharedRateLimitingStrategy::default(),
+            max_retries: DEFAULT_MAX_RETIRES,
+        }
+    }
+}
+
+impl FromStr for RateLimitingStrategy {
+    type Err = anyhow::Error;
+
+    fn from_str(config: &str) -> Result<Self> {
+        let mut parts = config.split(',');
+
+        let shared_config = parts.by_ref().take(3).collect::<Vec<_>>().join(",");
+        let inner = SharedRateLimitingStrategy::from_str(&shared_config)?;
+
+        let default_max_retries_str = DEFAULT_MAX_RETIRES.to_string();
+        let max_retries = parts.next().unwrap_or(&default_max_retries_str);
+        let max_retries = max_retries.parse().context("parsing max_retries")?;
+
+        ensure!(
+            parts.next().is_none(),
+            "extraneous rate limiting parameters"
+        );
+
+        Ok(Self { inner, max_retries })
+    }
+}
+
+#[derive(Error, Debug, Clone, Default, PartialEq)]
+pub enum RateLimiterError {
+    #[default]
+    #[error("rate limited")]
+    RateLimited,
+}
+
+impl RateLimiter {
+    pub fn new(strategy: RateLimitingStrategy, name: String) -> Self {
+        Self {
+            inner: SharedRateLimiter::from_strategy(strategy.inner, name),
+            max_retries: strategy.max_retries,
+        }
+    }
+
+    pub async fn execute<T>(
+        &self,
+        task: impl Future<Output = T>,
+        requires_back_off: impl Fn(&T) -> bool,
+    ) -> Result<T, RateLimiterError> {
+        self.inner
+            .execute(task, requires_back_off)
+            .await
+            .map_err(|err| match err {
+                SharedRateLimiterError::RateLimited => RateLimiterError::RateLimited,
+            })
+    }
+
+    pub async fn execute_with_retries<T, F, Fut>(
```

Comment:
> if a single order dex solver got rate limited it's fine for it to skip an order
Considering this, I see the following solution: if an order is skipped due to rate limiting, we wait until the cool-down ends and then continue with the following orders. That means the retry logic won't even be required. The only thing that needs to be added is a blocking mechanism for the thread to cool down.
This could be implemented in the shared rate limiter with a separate method like `execute_with_cooldown` or the service business logic itself.
Suggestion by Martin:
```
    let response = execute().await;
    if response.rate_limited() {
        wait_until_ready()
    }
```
Let me know if that makes sense.

Reply:
I wouldn't make the rate limiter block from giving us a response until the cooldown period is reached (this seems like the worst of both worlds, the initial request is slow only to tell us that we are rate limited).

If you feel strongly about delaying the next request (instead of the rate limiter just rejecting all subsequent requests as it's done for current rate limits), I think a retry is better. I'd just 
- implement it in the shared component
- have the caller specify the number of retries (rather than it being a top level struct parameter)

Reply:
I can explain why I decided to go with the retry: Don't really see any sense in returning `RateLimited` error for all the subsequent orders, meaning the whole auction will be skipped since the iteration over all the orders is indeed much faster than the cooldown period. 

Instead, we can give it a chance to cool down and try again within the original deadline.

Correct me if my assumption is wrong.

Reply:
>  the initial request is slow only to tell us that we are rate limited

My previous suggestion was wrong actually.
> if an order is skipped due to rate limiting, we wait until the cool-down ends and then continue with the following orders.

Instead, I wanted to do the opposite. If the first request rate is limited, then the following one waits once the cool down is over.